systemFlake.nix

```diff
@@ -25,25 +25,29 @@
 }@args:
 
 let
-  evalHostArgs =
-    { channelName ? "nixpkgs"
+  validateHostAttrs = {
+      channelName ? null
+    , system ? null
     , modules ? []
-    , system ? defaultSystem
     , extraArgs ? {}
-    , ...
-    }: defaultHostAttrs
-      // { 
-        inherit channelName system; 
-        modules = sharedModules ++ modules;
-        extraArgs = sharedExtraArgs // extraArgs;
-      };
+  } @ args: { inherits channelName system modules extraArgs; };
+
+  mergeHostAttrs = defaultAttrs: attrs: let
+    defaultTo = lhs: rhs: if rhs != null then rhs else lhs;
+  in
+  {
+    channelName = defaultTo (defaultTo "nixpkgs" defaultAttrs.channelName) attrs.channelName;
+    system = defaultTo (defaultTo defaultSystem defaultAttrs.system) attrs.system; # replace deaultSystem with x86_64-linux
+    modules = modules ++ defaultAttrs.modules ++ sharedModules;
+    extraArgs = extraArgs // defaultAttrs.extraArgs // sharedExtraArgs;
+  };
```

Comment:
Your again implementing basic nix language features like `?` and `//` when its not necessary. Combining into one function makes things simpler.

Reply:
> Combining into one function makes things simpler.

That would necessarily mean mixing the preprocessing of what's _externally_ fed to the API and the actual business logic.

I can't conform such to the idea of contract validator being _separate_ from logic.